cmd/smashwords-downloader/main.go

```diff
@@ -98,36 +128,316 @@ func scrapeBookList(pageId int, dataDir string) {
-	listCollector.Visit(westernRomanceURL)
+	smashwordsCategoryURL := fmt.Sprintf("https://%s/books/category/%d/downloads/0/free/any/%d", smashWordsURL, urlID, pageId)
+	listCollector.Visit(smashwordsCategoryURL)
 }
 
-func getDataDir() string {
-	dataDirPtr := flag.String("data_dir", "./data", "directory that the book files will download to")
+func main() {
+	//flags used: -url is the url to scrape,
+	// -data_dir is the directory to save the files to
+	dataDirPtr := flag.String("data_dir", "./data",
+		"directory that the book files will download to")
+
+	urlIDPtr := flag.Int("id", 1245,
+		"The cooresponding ID for the smashswords url you want to scrape"+
+			" (in https://www.smashwords.com/books/category/1245)")
+
+	itemsPerPagePtr := flag.Int("pageitems", 20,
+		"The number of items per page on the smashwords list page")
+
+	pagesPtr := flag.Int("pages", 7,
+		"The number of pages to scrape")
+
+	textFormatPtr := flag.String("format", "txt",
+		"The format of the book to download. Options are 'txt' or 'epub'")
+
+	overwriteSourcePtr := flag.Bool("overwriteSource", true,
+		"Save the original file after converting it to the desired format")
 	flag.Parse()
 
-	return *dataDirPtr
-}
+	totalBooks := *itemsPerPagePtr * *pagesPtr
 
-func main() {
-	dataDir := getDataDir()
+	//log the flag parameters out to console
+	log.Printf("Scraping %d pages of %d items, (total is %d) each from smashwords url %d.\n", *pagesPtr, *itemsPerPagePtr, totalBooks, *urlIDPtr)
+	log.Printf("Selected format is %s.\n", *textFormatPtr)
+	log.Printf("Saving files to %s folder.\n", *dataDirPtr)
 
+	// Create a wait group to wait for all the goroutines to finish
 	wg := new(sync.WaitGroup)
 
 	// Each list page only shows `bookListSize` books so scrape each one in parallel
-	for i := 0; i <= maxPageId; i = i + bookListSize {
+	for i := 0; i < (totalBooks); i = i + *itemsPerPagePtr {
 		wg.Add(1)
 		go func(pageId int) {
 			defer wg.Done()
-			scrapeBookList(pageId, dataDir)
+			scrapeBookList(pageId, *dataDirPtr, *urlIDPtr, *textFormatPtr)
 		}(i)
 	}
 
 	wg.Wait()
+
+	//convert epub to txt if needed
+	if *textFormatPtr == "epub" {
+
+		ConvertEpubGo(*dataDirPtr, *overwriteSourcePtr)
+	}
+}
+
+//A lot of the actual parsing is done with this repo: https://github.com/taylorskalyo/goreader
+func ConvertEpubGo(inputdir string, overwriteSource bool) {
+	//get all files in directory
+	files, err := ioutil.ReadDir(inputdir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//we time the parsing
+	start := time.Now()
+
+	//we count the number of characters
+	charCount := 0
+
+	//for each file, if it is an epub, convert it to txt
+	for _, file := range files {
+		if strings.HasSuffix(file.Name(), ".epub") {
+			filepath := inputdir + "/" + file.Name()
+
+			//We use the goreader library to parse the epub
+			rc, err := epub.OpenReader(filepath)
+			if err != nil {
+				panic(err)
+			}
+			defer rc.Close()
```

Comment:
didn't know that but that makes sense actually, since defer closes it after the entire function executes but depending on how many files are open, it may have just a giant stack of deferred files open.